R/aggregate_by_matrix.R

```diff
@@ -0,0 +1,37 @@
+#' @rdname aggregate
+#'
+#' @export
+colMeansMat <- function(x, MAT)
+  colSumsMat(x, MAT)/colSums(MAT)
+
+#' @rdname aggregate
+#'
+#' @export
+colSumsMat <- function(x, MAT)
+  t(MAT) %*% x
+
+
+##' @param MAT An adjacency matrix that defines what features with
+##'     `nrow(MAT) == nrow(x)`
+##'
+##' @return [aggregate_by_matrix()] returns a `matrix` (or `Matrix`)
+##'     of dimensions `ncol(MAT)` and `ncol(x), with `dimnames` equal
+##'     to `colnames(x)` and `rownames(MAT)`.
+##'
+##' @rdname aggregate
+##'
+##' @export
+aggregate_by_matrix <- function(x, MAT, FUN, ...) {
+    if (!inherits(x, "matrix"))
+        stop("'x' must be a matrix.")
+    if (!inherits(MAT, "Matrix") & !inherits(MAT, "matrix"))
+        stop("'MAT' must be a matrix.")
+    if (!identical(nrow(MAT), nrow(x)))
+        stop("nrow(MAT) must be identical to 'nrow(x).")
+    res <- do.call(FUN, list(x, MAT, ...))
+    if (!is.null(colnames(MAT)))
+        stopifnot(identical(colnames(MAT), rownames(res)))
```

Comment:
`stopifnot` is a slow function with multiple tests and an often unhelpful error message. That's why I would suggest the following:
```suggestion
    if (!is.null(colnames(MAT)) && !identical(colnames(MAT), rownames(res)))
        stop("The colum names of 'MAT' have to be identical to the row names of 'res'!")
```

Reply:
Thanks - I suppose I just like writing `stopifnot()` :-)